src/invocation/ClusterService.ts

```diff
@@ -14,317 +14,283 @@
-            this.ownerConnection = null;
-            this.connectToCluster().catch(this.client.shutdown.bind(this.client));
+    public handleMembersViewEvent(memberListVersion: number, memberInfos: MemberInfo[]): void {
+        if (this.memberListSnapshot === EMPTY_SNAPSHOT) {
+            this.applyInitialState(memberListVersion, memberInfos);
+            this.initialListFetched.resolve();
+            return;
         }
-    }
 
-    private onHeartbeatStopped(connection: ClientConnection): void {
-        this.logger.warn('ClusterService', connection.toString() + ' stopped heartbeating.');
-        if (connection.isAuthenticatedAsOwner()) {
-            this.client.getConnectionManager().destroyConnection(connection.getAddress());
+        let events = new Array<MembershipEvent>();
+        if (memberListVersion >= this.memberListSnapshot.version) {
+            const prevMembers = Array.from(this.memberListSnapshot.members.values());
+            const snapshot = this.createSnapshot(memberListVersion, memberInfos);
+            this.memberListSnapshot = snapshot;
+            const currentMembers = Array.from(snapshot.members.values());
+            events = this.detectMembershipEvents(prevMembers, currentMembers);
         }
+        this.fireEvents(events);
     }
 
-    private tryConnectingToAddresses(index: number, remainingAttemptLimit: number,
-                                     attemptPeriod: number, cause?: Error): Promise<void> {
-        this.logger.debug('ClusterService', 'Trying to connect to addresses, remaining attempt limit: ' + remainingAttemptLimit
-            + ', attempt period: ' + attemptPeriod);
-        if (this.knownAddresses.length <= index) {
-            remainingAttemptLimit = remainingAttemptLimit - 1;
-            if (remainingAttemptLimit === 0) {
-                const errorMessage = 'Unable to connect to any of the following addresses: ' +
-                    this.knownAddresses.map((element: Address) => {
-                        return element.toString();
-                    }).join(', ');
-                this.logger.debug('ClusterService', errorMessage);
-                const error = new IllegalStateError(errorMessage, cause);
-                return Promise.reject(error);
-            } else {
-                const deferred = DeferredPromise<void>();
-                setTimeout(
-                    () => {
-                        this.tryConnectingToAddresses(0, remainingAttemptLimit, attemptPeriod).then(() => {
-                            deferred.resolve();
-                        }).catch((e) => {
-                            deferred.reject(e);
-                        });
-                    },
-                    attemptPeriod,
-                );
-                return deferred.promise;
+    private fireEvents(events: MembershipEvent[]): void {
+        for (const event of events) {
+            for (const listener of Array.from(this.listeners.values())) {
```

Comment:
We could avoid litter arrays by introducing `listenersList` field with pre-calculated array of listeners. The idea is that same as for the comment on `MemberListSnapshot.membersList`.

Reply:
I used `forEach` for the listeners as you suggested